#####  **Import Required Libraries**

In [2]:
import pandas as pd

#####  **Read the Input Data**

In [3]:
# Load the dataset
df = pd.read_csv('../resources/data/input/Customer Churn.csv')

# Display the first 5 rows of the dataset
df.head()

call_failure  complains  subscription_length  charge_amount  \
0             8          0                   38              0   
1             0          0                   39              0   
2            10          0                   37              0   
3            10          0                   38              0   
4             3          0                   38              0   

   seconds_of_use  frequency_of_use  frequency_of_SMS  \
0            4370                71                 5   
1             318                 5                 7   
2            2453                60               359   
3            4198                66                 1   
4            2393                58                 2   

   distinct_called_numbers  age_group  tariff_plan  status  age  \
0                       17          3            1       1   30   
1                        4          2            1       2   25   
2                       24          3            1       1   30   
3                       35          1            1       1   15   
4                       33          1            1       1   15   

   customer_value  churn  
0         197.640      0  
1          46.035      0  
2        1536.520      0  
3         240.020      0  
4         145.805      0

#####  **Check Missing Values**

In [4]:
# Check the shape of the dataset
print(f"Dataset Shape: {df.shape}")

# Check for missing values
print(df.isnull().sum())

# Check data types of each column
df.dtypes

Dataset Shape: (3150, 14)
call_failure               0
complains                  0
subscription_length        0
charge_amount              0
seconds_of_use             0
frequency_of_use           0
frequency_of_SMS           0
distinct_called_numbers    0
age_group                  0
tariff_plan                0
status                     0
age                        0
customer_value             0
churn                      0
dtype: int64


call_failure                 int64
complains                    int64
subscription_length          int64
charge_amount                int64
seconds_of_use               int64
frequency_of_use             int64
frequency_of_SMS             int64
distinct_called_numbers      int64
age_group                    int64
tariff_plan                  int64
status                       int64
age                          int64
customer_value             float64
churn                        int64
dtype: object

##### **Check Duplicates**

In [5]:
# Check for duplicate rows
print(f"Number of duplicate rows: {df.duplicated().sum()}")

Number of duplicate rows: 300


In [6]:
# Drop duplicate rows
df = df.drop_duplicates()

# Verify if duplicates are removed
print(f"Number of duplicate rows after cleaning: {df.duplicated().sum()}")

Number of duplicate rows after cleaning: 0


In [7]:
print(f"Dataset Shape after dropping duplicates: {df.shape}")

Dataset Shape after dropping duplicates: (2850, 14)


In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 2850 entries, 0 to 3131
Data columns (total 14 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   call_failure             2850 non-null   int64  
 1   complains                2850 non-null   int64  
 2   subscription_length      2850 non-null   int64  
 3   charge_amount            2850 non-null   int64  
 4   seconds_of_use           2850 non-null   int64  
 5   frequency_of_use         2850 non-null   int64  
 6   frequency_of_SMS         2850 non-null   int64  
 7   distinct_called_numbers  2850 non-null   int64  
 8   age_group                2850 non-null   int64  
 9   tariff_plan              2850 non-null   int64  
 10  status                   2850 non-null   int64  
 11  age                      2850 non-null   int64  
 12  customer_value           2850 non-null   float64
 13  churn                    2850 non-null   int64  
dtypes: float64(1), int64(13)
memo

##### **Check invalid values**

| Attribute                | Description                                                                                     | Validity Considerations                                                                                     |
|--------------------------|-------------------------------------------------------------------------------------------------|-------------------------------------------------------------------------------------------------------------|
| **Call Failures**        | Number of call failures. Zero is perfectly valid (no failures). Negative values should be investigated/corrected. | Valid: 0; Invalid: Negative values.                                                                         |
| **Complains**            | Binary (0: No complaint, 1: Complaint). Zero complaints are valid, but negative values are not. | Valid: 0, 1; Invalid: Negative values.                                                                       |
| **Subscription Length**  | Total months of subscription. Zero could represent a new subscriber or trial period. Negative values don't make sense. | Valid: 0; Invalid: Negative values.                                                                          |
| **Charge Amount**        | Amount charged to the customer. Zero could indicate a free trial or promotional period. Negative values may represent refunds or credits. | Valid: 0; Invalid: Negative values (depending on context).                                                  |
| **Seconds of Use**       | Total seconds of calls. Zero is valid (no usage), negative values are not.                     | Valid: 0; Invalid: Negative values.                                                                          |
| **Frequency of Use**     | Total number of calls. Zero is valid, negative values are not.                                  | Valid: 0; Invalid: Negative values.                                                                          |
| **Frequency of SMS**     | Total number of text messages. Zero is valid, negative values are not.                         | Valid: 0; Invalid: Negative values.                                                                          |
| **Distinct Called Numbers** | Total number of distinct phone calls. Zero is valid (no calls made), negative values are not. | Valid: 0; Invalid: Negative values.                                                                          |
| **Age Group**            | Ordinal attribute (1: younger age, 5: older age). Age group should never be negative. Zero might be a valid category (e.g., infants). | Invalid: Negative values; Valid: 0 (depends on context).                                                     |
| **Tariff Plan**          | Binary (1: Pay as you go, 2: Contractual). Zero might or might not be valid depending on the coding scheme. Negative values are likely invalid. | Valid: Depends on context; Invalid: Negative values.                                                         |
| **Status**               | Binary (1: Active, 2: Non-active). Validity of zero or negative values depends on how these statuses are coded. | Valid: Depends on coding scheme; Invalid: Negative values.                                                  |
| **Churn**                | Binary (1: Churn, 0: Non-churn) - Class label. Zero and one are the only expected values. Negative values would be invalid. | Valid: 0, 1; Invalid: Negative values.                                                                       |
| **Customer Value**       | The calculated value of the customer. Zero could be valid (e.g., new customer). Negative values might represent potential loss or churn risk. | Valid: 0; Invalid: Negative values (depends on specific definition).                                       |

In [9]:
df.describe()

call_failure    complains  subscription_length  charge_amount  \
count   2850.000000  2850.000000          2850.000000    2850.000000   
mean       7.802456     0.080702            32.452982       0.974737   
std        7.326172     0.272424             8.723075       1.550618   
min        0.000000     0.000000             3.000000       0.000000   
25%        1.000000     0.000000            29.000000       0.000000   
50%        6.000000     0.000000            35.000000       0.000000   
75%       12.000000     0.000000            38.000000       2.000000   
max       36.000000     1.000000            47.000000      10.000000   

       seconds_of_use  frequency_of_use  frequency_of_SMS  \
count     2850.000000       2850.000000       2850.000000   
mean      4534.243158         70.484912         73.789825   
std       4199.712303         57.401512        112.062397   
min          0.000000          0.000000          0.000000   
25%       1458.750000         28.000000          7.000000   
50%       3041.000000         54.500000         22.000000   
75%       6500.000000         96.000000         88.000000   
max      17090.000000        255.000000        522.000000   

       distinct_called_numbers    age_group  tariff_plan       status  \
count              2850.000000  2850.000000  2850.000000  2850.000000   
mean                 23.870526     2.835088     1.080351     1.240000   
std                  17.193929     0.893503     0.271883     0.427158   
min                   0.000000     1.000000     1.000000     1.000000   
25%                  11.000000     2.000000     1.000000     1.000000   
50%                  21.000000     3.000000     1.000000     1.000000   
75%                  34.000000     3.000000     1.000000     1.000000   
max                  97.000000     5.000000     2.000000     2.000000   

               age  customer_value        churn  
count  2850.000000     2850.000000  2850.000000  
mean     31.077193      474.990367     0.156491  
std       8.861934      514.442198     0.363384  
min      15.000000        0.000000     0.000000  
25%      25.000000      117.527500     0.000000  
50%      30.000000      232.520000     0.000000  
75%      30.000000      790.080000     0.000000  
max      55.000000     2165.280000     1.000000

The following table presents an analysis of various fields in the dataset, specifically aimed at identifying invalid values as part of the data cleaning process. Each feature has been evaluated for its minimum (Min) and maximum (Max) observed values, with a particular focus on highlighting any potential data quality issues.

| Feature               | Min | Max | Notes                                                                  |
|-----------------------|-----|-----|------------------------------------------------------------------------|
| call_failure          | 0   | 36  | No negative values                                                    |
| complains             | 0   | 1   | Binary (0 or 1)                                                        |
| subscription_length   | 3   | 47  | No negative values                                                    |
| charge_amount         | 0   | 10  | No negative values (though negative values *might* represent refunds) |
| seconds_of_use        | 0   | 17090| No negative values                                                    |
| frequency_of_use      | 0   | 255 | No negative values                                                    |
| frequency_of_SMS      | 0   | 522 | No negative values                                                    |
| distinct_called_numbers| 0   | 97  | No negative values                                                    |
| age_group             | 1   | 5   | Categorical                                                           |
| tariff_plan           | 1   | 2   | Categorical                                                           |
| status                | 1   | 2   | Categorical                                                           |
| age                   | 15  | 55  | No negative values                                                    |
| customer_value        | 0   | 2165| No negative values                                                    |
| churn                 | 0   | 1   | Binary (0 or 1)                                                        |

#####  **Write Output**

In [10]:
# Save the cleaned dataset
df.to_csv('../resources/data/output/customer_churn_cleaned.csv', index=False)